server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java

```diff
@@ -267,4 +277,128 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
             }
         });
     }
+
+    void setWeightForDecommissionedZone(List<String> zones, ActionListener<ClusterPutWRRWeightsResponse> listener) {
+        ClusterState clusterState = clusterService.getClusterApplierService().state();
+
+        DecommissionAttributeMetadata decommissionAttributeMetadata = clusterState.metadata().custom(DecommissionAttributeMetadata.TYPE);
+        assert decommissionAttributeMetadata.status().equals(DecommissionStatus.WEIGH_AWAY)
+            : "unexpected status encountered while decommissioning nodes";
+        DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+
+        Map<String, String> weights = new HashMap<>();
+        zones.forEach(zone -> {
+            if (zone.equalsIgnoreCase(decommissionAttribute.attributeValue())) {
+                weights.put(zone, "0");
+            } else {
+                weights.put(zone, "1");
+            }
+        });
+
+        // WRR API will validate invalid weights
+        final ClusterPutWRRWeightsRequest clusterWeightRequest = new ClusterPutWRRWeightsRequest();
+        clusterWeightRequest.attributeName(decommissionAttribute.attributeName());
+        clusterWeightRequest.setWRRWeight(weights);
+
+        transportService.sendRequest(
+            transportService.getLocalNode(),
+            ClusterPutWRRWeightsAction.NAME,
+            clusterWeightRequest,
+            new TransportResponseHandler<ClusterPutWRRWeightsResponse>() {
+                @Override
+                public void handleResponse(ClusterPutWRRWeightsResponse response) {
+                    logger.info("Weights are successfully set.");
+                    listener.onResponse(response);
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    // Logging warn message on failure. Should we do Retry? If weights are not set should we fail?
+                    logger.warn("Exception occurred while setting weights.Exception Messages - [{}]", exp.unwrapCause().getMessage());
+                    listener.onFailure(exp);
```

Comment:
How are we handling failures?

Reply:
The listener will call clearVotingConfig with appropriate parameter

```
                decommissionController.setWeight(awarenessValues, new ActionListener<>() {
                    @Override
                    public void onResponse(ClusterPutWeightedRoutingResponse response) {
                        scheduleNodesDecommissionOnTimeout(state, timeOutForNodeDraining);
                    }

                    @Override
                    public void onFailure(Exception e) {
                        clearVotingConfigExclusionAndUpdateStatus(false, false);
                    }
                });
```